lib/core/buffer.hpp

```diff
@@ -41,30 +35,309 @@
+     * calling @p mark_frame_empty(), because it could be dereferenced and returned
+     * to the metadata pool after that call.
+     * If you are adding a new metadata object, please *also* call
+     * @c allocate_new_metadata_object() before asking for the metadata object.
+     *
+     * @param[in] frame_id The frame to return the metadata for.
+     * @returns A pointer to the metadata_container
+     */
+    metadataContainer* get_metadata_container(int frame_id);
+
+    /// The main lock for frame state management
+    std::recursive_mutex mutex;
+
+    /// The number of frames kept by this object
+    int num_frames;
+
+    /// Should we shut down (stop returning new frames)?
+    bool shutdown_signal;
+
+    /// The list of consumer names registered to this buffer
+    std::map<std::string, StageInfo> consumers;
+
+    /// The list of producer names registered to this buffer
+    std::map<std::string, StageInfo> producers;
+
+    /// The name of the buffer for use in debug messages.
+    std::string buffer_name;
+
+    /// The type of the buffer for use in writing data.
+    std::string buffer_type;
+
+    /// The pool of info objects
+    metadataPool* metadata_pool;
+
+    /// Array of buffer info objects, for tracking information about each buffer.
+    std::vector<metadataContainer*> metadata;
+
+protected:
+    /// The condition variable for calls to @c wait_for_full_buffer
+    std::condition_variable_any full_cond;
+
+    /// The condition variable for calls to @c wait_for_empty_buffer
+    std::condition_variable_any empty_cond;
+};
+
+/**
+ * @brief A buffer to manage the signalling between stages when those
+ * stages want to communicate data using a ring buffer.
+ *
+ * This class is a little strange in that it doesn't actually allocate
+ * any memory!  It was created to connect GPU pipelines that need to
+ * communicate via a ring buffer in GPU memory.  This class knows how
+ * big the buffer is and controls read and write access for the
+ * consumers and producers, but does not actually own the memory.
+ *
+ * Producers and consumers can produce or consume different numbers of
+ * elements, so this is a helpful way of connecting stages that
+ * operate at different cadences.
+ *
+ * In the yaml config file, one of these can be created by setting the
+ * @c kotekan_buffer value to @c "ring".
+ *
+ * @conf ring_buffer_size: the number of elements in the ring buffer
+ * @conf metadata_pool The name of the metadata pool to associate with the buffer
+ */
+class RingBuffer : public GenericBuffer {
+public:
+    /**
+     * @brief Build a new RingBuffer for connecting kotekan stages
+     * that want to produce & consume data with different sizes or at
+     * different cadences.
+     *
+     * Currently, only a single producer and single consumer are supported!
+     *
+     * @param ring_size: the number of elements in the ring buffer to be managed
```

Comment:
Do you see a case where elements != bytes?    If not, I'd be tempted to make this bytes to make it easier to follow.

Reply:
I was thinking it could be convenient to handle elements != bytes, because the RingBuffer could log the logical units rather than the pointers and bytes.  But maybe that's too much extra complexity for not enough payoff!